File: Multi-Way Trees/2-3-4-tree-implementation.c
/* File: 2-3-4-tree-implementation.c
   Program implementing basic 2-3-4 Tree operations
*/

#include <stdio.h>
#include <stdlib.h>

#define MAX 3
#define MIN 1

struct node {
    int n;
    int keys[MAX+1];
    struct node *child[MAX+2];
};

struct node *root = NULL;

struct node* createNode(int key)
{
    struct node *newNode = (struct node*)malloc(sizeof(struct node));

    newNode->n = 1;
    newNode->keys[0] = key;

    for(int i=0;i<5;i++)
        newNode->child[i] = NULL;

    return newNode;
}

void traverse(struct node *ptr)
{
    if(ptr)
    {
        int i;

        for(i=0;i<ptr->n;i++)
        {
            traverse(ptr->child[i]);
            printf("%d ",ptr->keys[i]);
        }

        traverse(ptr->child[i]);
    }
}

struct node* search(struct node *ptr, int key)
{
    int i=0;

    if(ptr==NULL)
        return NULL;

    while(i<ptr->n && key>ptr->keys[i])
        i++;

    if(i<ptr->n && key==ptr->keys[i])
        return ptr;

    return search(ptr->child[i],key);
}

void insert(int key)
{
    if(root==NULL)
    {
        root = createNode(key);
        return;
    }

    if(root->n < MAX)
    {
        root->keys[root->n++] = key;
    }
}

int main()
{
    insert(10);
    insert(20);
    insert(5);
    insert(6);
    insert(12);

    printf("Traversal of 2-3-4 Tree:\n");

    traverse(root);

    int key=12;

    if(search(root,key))
        printf("\nKey found");
    else
        printf("\nKey not found");

    return 0;
}

File: Multi-Way Trees/2-3-4-tree-implementation.java
// File: 2-3-4-tree-implementation.java
// Program implementing 2-3-4 Tree operations

class Node {

    int keys[] = new int[3];
    Node children[] = new Node[4];
    int numKeys = 0;
    boolean leaf = true;
}

public class TwoThreeFourTree {

    Node root = new Node();

    void traverse(Node node)
    {
        if(node==null)
            return;

        int i;

        for(i=0;i<node.numKeys;i++)
        {
            if(!node.leaf)
                traverse(node.children[i]);

            System.out.print(node.keys[i]+" ");
        }

        if(!node.leaf)
            traverse(node.children[i]);
    }

    Node search(Node node,int key)
    {
        int i=0;

        while(i<node.numKeys && key>node.keys[i])
            i++;

        if(i<node.numKeys && key==node.keys[i])
            return node;

        if(node.leaf)
            return null;

        return search(node.children[i],key);
    }

    void insert(int key)
    {
        if(root.numKeys<3)
        {
            root.keys[root.numKeys++] = key;
        }
    }

    public static void main(String args[])
    {
        TwoThreeFourTree tree = new TwoThreeFourTree();

        tree.insert(10);
        tree.insert(20);
        tree.insert(5);

        System.out.println("Traversal of 2-3-4 Tree:");

        tree.traverse(tree.root);
    }
}
